{
 "cells": [
  {
   "cell_type": "code",
   "id": "226559791cc55464",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:30:36.299980Z",
     "start_time": "2025-11-19T20:30:33.315486Z"
    }
   },
   "source": [
    "import os\n",
    "\n",
    "import decoupler as dc\n",
    "import numba\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "numba.set_num_threads(8)"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "id": "cc67e49ff57329fd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T17:48:32.234935Z",
     "start_time": "2025-11-19T17:48:26.568748Z"
    }
   },
   "source": [
    "# adata_file = \"data/scPerturbData/TianKampmann2021_CRISPRi.h5ad\"\n",
    "adata_file = \"data/scPerturbData/AdamsonWeissman2016_GSM2406681_10X010.h5ad\"\n",
    "adata = sc.read_h5ad(adata_file)\n",
    "adata"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 65337 × 32738\n",
       "    obs: 'perturbation', 'read count', 'UMI count', 'tissue_type', 'cell_line', 'cancer', 'disease', 'perturbation_type', 'celltype', 'organism', 'ncounts', 'ngenes', 'percent_mito', 'percent_ribo', 'nperts'\n",
       "    var: 'ensembl_id', 'ncounts', 'ncells'"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 99
  },
  {
   "cell_type": "code",
   "id": "b81cf7f46c386e28",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T17:49:00.738015Z",
     "start_time": "2025-11-19T17:49:00.705162Z"
    }
   },
   "source": [
    "adata.obs"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                   perturbation  read count  UMI count tissue_type cell_line  \\\n",
       "cell_barcode                                                                   \n",
       "AAACATACAAGATG   63(mod)_pBA580       282.0        8.0   cell_line      K562   \n",
       "AAACATACACCTAG      OST4_pDS353       331.0        7.0   cell_line      K562   \n",
       "AAACATACTTCCCG   SEC61A1_pDS031       285.0       10.0   cell_line      K562   \n",
       "AAACATTGAAACAG    EIF2B4_pDS491      1036.0       30.0   cell_line      K562   \n",
       "AAACATTGCAGCTA      SRPR_pDS482       863.0       25.0   cell_line      K562   \n",
       "...                         ...         ...        ...         ...       ...   \n",
       "TTTGCATGCTTTAC     STT3A_pDS011       476.0       17.0   cell_line      K562   \n",
       "TTTGCATGGAGGAC  ARHGAP22_pDS458       539.0       19.0   cell_line      K562   \n",
       "TTTGCATGTAGAGA   63(mod)_pBA580       647.0       35.0   cell_line      K562   \n",
       "TTTGCATGTCAAGC    KCTD16_pDS096        98.0        4.0   cell_line      K562   \n",
       "TTTGCATGTGGAGG    SAMM50_pDS156       427.0       14.0   cell_line      K562   \n",
       "\n",
       "                cancer                       disease perturbation_type  \\\n",
       "cell_barcode                                                             \n",
       "AAACATACAAGATG    True  chronic myelogenous leukemia            CRISPR   \n",
       "AAACATACACCTAG    True  chronic myelogenous leukemia            CRISPR   \n",
       "AAACATACTTCCCG    True  chronic myelogenous leukemia            CRISPR   \n",
       "AAACATTGAAACAG    True  chronic myelogenous leukemia            CRISPR   \n",
       "AAACATTGCAGCTA    True  chronic myelogenous leukemia            CRISPR   \n",
       "...                ...                           ...               ...   \n",
       "TTTGCATGCTTTAC    True  chronic myelogenous leukemia            CRISPR   \n",
       "TTTGCATGGAGGAC    True  chronic myelogenous leukemia            CRISPR   \n",
       "TTTGCATGTAGAGA    True  chronic myelogenous leukemia            CRISPR   \n",
       "TTTGCATGTCAAGC    True  chronic myelogenous leukemia            CRISPR   \n",
       "TTTGCATGTGGAGG    True  chronic myelogenous leukemia            CRISPR   \n",
       "\n",
       "                    celltype organism  ncounts  ngenes  percent_mito  \\\n",
       "cell_barcode                                                           \n",
       "AAACATACAAGATG  lymphoblasts    human   8866.0    2914      4.917663   \n",
       "AAACATACACCTAG  lymphoblasts    human  13785.0    3818      4.468626   \n",
       "AAACATACTTCCCG  lymphoblasts    human   7569.0    2616      5.060113   \n",
       "AAACATTGAAACAG  lymphoblasts    human  13834.0    3488      5.052769   \n",
       "AAACATTGCAGCTA  lymphoblasts    human  15507.0    3620      4.514091   \n",
       "...                      ...      ...      ...     ...           ...   \n",
       "TTTGCATGCTTTAC  lymphoblasts    human  14524.0    3356      5.996971   \n",
       "TTTGCATGGAGGAC  lymphoblasts    human  11685.0    2961      4.612751   \n",
       "TTTGCATGTAGAGA  lymphoblasts    human  16610.0    3473      7.242625   \n",
       "TTTGCATGTCAAGC  lymphoblasts    human  14473.0    3431      7.296345   \n",
       "TTTGCATGTGGAGG  lymphoblasts    human  16843.0    3719      5.919373   \n",
       "\n",
       "                percent_ribo  nperts  \n",
       "cell_barcode                          \n",
       "AAACATACAAGATG     21.306112       2  \n",
       "AAACATACACCTAG     19.492201       2  \n",
       "AAACATACTTCCCG     23.199894       2  \n",
       "AAACATTGAAACAG     28.733555       2  \n",
       "AAACATTGCAGCTA     26.729864       2  \n",
       "...                      ...     ...  \n",
       "TTTGCATGCTTTAC     22.679703       2  \n",
       "TTTGCATGGAGGAC     26.983313       2  \n",
       "TTTGCATGTAGAGA     26.207104       2  \n",
       "TTTGCATGTCAAGC     20.755890       2  \n",
       "TTTGCATGTGGAGG     20.097370       2  \n",
       "\n",
       "[65337 rows x 15 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>perturbation</th>\n",
       "      <th>read count</th>\n",
       "      <th>UMI count</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cancer</th>\n",
       "      <th>disease</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>celltype</th>\n",
       "      <th>organism</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>nperts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAAGATG</th>\n",
       "      <td>63(mod)_pBA580</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>8866.0</td>\n",
       "      <td>2914</td>\n",
       "      <td>4.917663</td>\n",
       "      <td>21.306112</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACACCTAG</th>\n",
       "      <td>OST4_pDS353</td>\n",
       "      <td>331.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>13785.0</td>\n",
       "      <td>3818</td>\n",
       "      <td>4.468626</td>\n",
       "      <td>19.492201</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACTTCCCG</th>\n",
       "      <td>SEC61A1_pDS031</td>\n",
       "      <td>285.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>7569.0</td>\n",
       "      <td>2616</td>\n",
       "      <td>5.060113</td>\n",
       "      <td>23.199894</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAAACAG</th>\n",
       "      <td>EIF2B4_pDS491</td>\n",
       "      <td>1036.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>13834.0</td>\n",
       "      <td>3488</td>\n",
       "      <td>5.052769</td>\n",
       "      <td>28.733555</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGCAGCTA</th>\n",
       "      <td>SRPR_pDS482</td>\n",
       "      <td>863.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>15507.0</td>\n",
       "      <td>3620</td>\n",
       "      <td>4.514091</td>\n",
       "      <td>26.729864</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGCTTTAC</th>\n",
       "      <td>STT3A_pDS011</td>\n",
       "      <td>476.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>14524.0</td>\n",
       "      <td>3356</td>\n",
       "      <td>5.996971</td>\n",
       "      <td>22.679703</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGGAGGAC</th>\n",
       "      <td>ARHGAP22_pDS458</td>\n",
       "      <td>539.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>11685.0</td>\n",
       "      <td>2961</td>\n",
       "      <td>4.612751</td>\n",
       "      <td>26.983313</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGTAGAGA</th>\n",
       "      <td>63(mod)_pBA580</td>\n",
       "      <td>647.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>16610.0</td>\n",
       "      <td>3473</td>\n",
       "      <td>7.242625</td>\n",
       "      <td>26.207104</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGTCAAGC</th>\n",
       "      <td>KCTD16_pDS096</td>\n",
       "      <td>98.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>14473.0</td>\n",
       "      <td>3431</td>\n",
       "      <td>7.296345</td>\n",
       "      <td>20.755890</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGTGGAGG</th>\n",
       "      <td>SAMM50_pDS156</td>\n",
       "      <td>427.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>K562</td>\n",
       "      <td>True</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>lymphoblasts</td>\n",
       "      <td>human</td>\n",
       "      <td>16843.0</td>\n",
       "      <td>3719</td>\n",
       "      <td>5.919373</td>\n",
       "      <td>20.097370</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65337 rows × 15 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 100
  },
  {
   "cell_type": "code",
   "id": "bfacfa99f50de3ea",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T17:49:29.626661Z",
     "start_time": "2025-11-19T17:49:29.621142Z"
    }
   },
   "source": [
    "print(adata.var)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   ensembl_id  ncounts  ncells\n",
      "gene_symbol                                   \n",
      "MIR1302-10    ENSG00000243485     11.0      11\n",
      "FAM138A       ENSG00000237613      0.0       0\n",
      "OR4F5         ENSG00000186092      0.0       0\n",
      "RP11-34P13.7  ENSG00000238009      0.0       0\n",
      "RP11-34P13.8  ENSG00000239945     43.0      43\n",
      "...                       ...      ...     ...\n",
      "AC145205.1    ENSG00000215635      0.0       0\n",
      "BAGE5         ENSG00000268590      0.0       0\n",
      "CU459201.1    ENSG00000251180      0.0       0\n",
      "AC002321.2    ENSG00000215616      0.0       0\n",
      "AC002321.1    ENSG00000215611      0.0       0\n",
      "\n",
      "[32738 rows x 3 columns]\n"
     ]
    }
   ],
   "execution_count": 101
  },
  {
   "cell_type": "code",
   "id": "e40196f7cfc45548",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T17:49:31.640509Z",
     "start_time": "2025-11-19T17:49:31.605369Z"
    }
   },
   "source": [
    "adata.obs.describe()"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "         read count     UMI count       ncounts        ngenes  percent_mito  \\\n",
       "count  62724.000000  62724.000000  65337.000000  65337.000000  65337.000000   \n",
       "mean     638.174734     24.956954  15915.297852   3639.789813      5.327863   \n",
       "std      567.115265     22.359175   6349.257812    803.389180      2.475392   \n",
       "min        1.000000      1.000000   3179.000000    462.000000      0.020105   \n",
       "25%      305.000000     12.000000  11669.000000   3148.000000      4.290833   \n",
       "50%      541.000000     21.000000  15355.000000   3690.000000      5.203821   \n",
       "75%      858.000000     32.000000  19572.000000   4192.000000      6.187406   \n",
       "max    46452.000000   1557.000000  67075.000000   6859.000000     82.174286   \n",
       "\n",
       "       percent_ribo        nperts  \n",
       "count  65337.000000  65337.000000  \n",
       "mean      24.905811      1.918469  \n",
       "std        3.947483      0.393537  \n",
       "min        2.047059      0.000000  \n",
       "25%       22.444645      2.000000  \n",
       "50%       24.729574      2.000000  \n",
       "75%       27.206667      2.000000  \n",
       "max       56.376625      2.000000  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read count</th>\n",
       "      <th>UMI count</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>nperts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>62724.000000</td>\n",
       "      <td>62724.000000</td>\n",
       "      <td>65337.000000</td>\n",
       "      <td>65337.000000</td>\n",
       "      <td>65337.000000</td>\n",
       "      <td>65337.000000</td>\n",
       "      <td>65337.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>638.174734</td>\n",
       "      <td>24.956954</td>\n",
       "      <td>15915.297852</td>\n",
       "      <td>3639.789813</td>\n",
       "      <td>5.327863</td>\n",
       "      <td>24.905811</td>\n",
       "      <td>1.918469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>567.115265</td>\n",
       "      <td>22.359175</td>\n",
       "      <td>6349.257812</td>\n",
       "      <td>803.389180</td>\n",
       "      <td>2.475392</td>\n",
       "      <td>3.947483</td>\n",
       "      <td>0.393537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3179.000000</td>\n",
       "      <td>462.000000</td>\n",
       "      <td>0.020105</td>\n",
       "      <td>2.047059</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>305.000000</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>11669.000000</td>\n",
       "      <td>3148.000000</td>\n",
       "      <td>4.290833</td>\n",
       "      <td>22.444645</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>541.000000</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>15355.000000</td>\n",
       "      <td>3690.000000</td>\n",
       "      <td>5.203821</td>\n",
       "      <td>24.729574</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>858.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>19572.000000</td>\n",
       "      <td>4192.000000</td>\n",
       "      <td>6.187406</td>\n",
       "      <td>27.206667</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>46452.000000</td>\n",
       "      <td>1557.000000</td>\n",
       "      <td>67075.000000</td>\n",
       "      <td>6859.000000</td>\n",
       "      <td>82.174286</td>\n",
       "      <td>56.376625</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 102
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "43683f9286c62590",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-12T19:32:01.542817Z",
     "start_time": "2025-11-12T19:31:58.051341Z"
    }
   },
   "outputs": [],
   "source": [
    "# identify controls (heuristic: explicit \"control\" in perturbation or guide names or nperts==0)\n",
    "ctrl_mask = (adata.obs['perturbation'].str.contains('control', case=False, na=False)) | \\\n",
    "            (adata.obs['guide_id'].str.contains('non-target', case=False, na=False)) | \\\n",
    "            (adata.obs['nperts'] == 0)\n",
    "adata.obs['is_control'] = ctrl_mask\n",
    "\n",
    "# create a per-cell label: target gene name or \"control\"\n",
    "adata.obs['ptb_label'] = adata.obs['perturbation'].astype(str).replace('nan', 'unknown')\n",
    "adata.obs.loc[adata.obs['is_control'], 'ptb_label'] = 'control'\n",
    "\n",
    "# Option: simplify labels to only TFs you care about\n",
    "# tf_list = [...]  # your TF list\n",
    "# adata.obs['ptb_label'] = adata.obs['ptb_label'].where(adata.obs['ptb_label'].isin(tf_list), other='other')\n",
    "\n",
    "# Save labels for evaluation\n",
    "labels = adata.obs['ptb_label']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8a3340c0491e385c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-12T20:00:10.346551Z",
     "start_time": "2025-11-12T20:00:10.339426Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels.shape:  (32300,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AAACCCAAGGGTTGCA    CREBBP\n",
       "AAACCCAAGGTAGCCA    SH3RF1\n",
       "AAACCCACAAACTAGA      TAF1\n",
       "AAACCCACAGAATTCC     UBTD2\n",
       "AAACCCAGTAGCGTTT    FRMD4A\n",
       "                     ...  \n",
       "TTTGTTGCAGGTCTCG    FAM57B\n",
       "TTTGTTGCAGTTGCGC     XRCC1\n",
       "TTTGTTGGTCCCGGTA     RPL14\n",
       "TTTGTTGGTTGATGTC     PMPCA\n",
       "TTTGTTGTCTTACTGT       HTT\n",
       "Name: ptb_label, Length: 32300, dtype: object"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"labels.shape: \", labels.shape)\n",
    "labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3de843c4742f69e0",
   "metadata": {},
   "source": "### Perturbation data filtering, cleanup for faster processing"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37248f04a1bbf2b1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-17T20:38:45.724314Z",
     "start_time": "2025-11-17T19:24:08.037748Z"
    },
    "jupyter": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file:  data/scPerturbData/ReplogleWeissman2022_K562_essential.h5ad\n",
      "Original Shape:  (310385, 8563)\n"
     ]
    }
   ],
   "source": [
    "adata_files = [\n",
    "    \"data/scPerturbData/TianKampmann2021_CRISPRi.h5ad\",  # (32300,  33538)   After Filtering   (31407,  21903)\n",
    "    \"data/scPerturbData/TianKampmann2021_CRISPRa.h5ad\",  # (21193,  33538)   After Filtering   (19758,  20404)\n",
    "    \"data/scPerturbData/ReplogleWeissman2022_rpe1.h5ad\",  # (247914, 8749 )   After Filtering   (247914, 8749 )\n",
    "    \"data/scPerturbData/NormanWeissman2019_filtered.h5ad\",  # (111445, 33694)   After Filtering   (111391, 20265)\n",
    "    \"data/scPerturbData/FrangiehIzar2021_RNA.h5ad\",  # (218331, 23712)   After Filtering   (218027, 23593)\n",
    "    ### \"data/scPerturbData/ReplogleWeissman2022_K562_essential.h5ad\",  # (310385, 8563) Too large to process here\n",
    "    \"data/scPerturbData/ShifrutMarson2018_kt_filtered.h5ad\",  # () After Filtering (52236, 33694)\n",
    "    \"data/scPerturbData/AdamsonWeissman2016_GSM2406675_10X001.h5ad\",  # (5768, 35635) After Filtering (3999, 12487)\n",
    "    \"data/scPerturbData/AdamsonWeissman2016_GSM2406681_10X010.h5ad\",  # (65337, 32738) After Filtering (59171, 18276)\n",
    "]\n",
    "\n",
    "for adata_file in adata_files:\n",
    "    print(\"Processing file: \", adata_file)\n",
    "    adata = sc.read_h5ad(adata_file)\n",
    "\n",
    "    print(\"Original Shape: \", adata.shape)\n",
    "    n_cells, n_genes = adata.shape\n",
    "\n",
    "    # Filter low-quality genes and cells\n",
    "    sc.pp.filter_cells(adata, min_genes=500)\n",
    "    sc.pp.filter_genes(adata, min_cells=10)\n",
    "    print(\"Shape after filtering: \", adata.shape)\n",
    "\n",
    "    # Mitochondrial/QC filter\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('MT-')  # assuming human data; use 'mt-' for mouse\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "    adata = adata[adata.obs.pct_counts_mt < 20, :]  # filter cells with >20% mitochondrial genes\n",
    "    print(\"Final Shape \", adata.shape)\n",
    "\n",
    "    # Normalize and log-transform\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    sc.pp.log1p(adata)\n",
    "    print(\"After Normalization and Log1p: \", adata.shape)\n",
    "\n",
    "    # Export adata after filtering\n",
    "    output_filtered_file = adata_file.replace('.h5ad', '_kt_filtered.h5ad')\n",
    "    adata.write_h5ad(output_filtered_file)\n",
    "    print(f\"Filtered data saved to {output_filtered_file}\")\n",
    "\n",
    "    print(\"*\" * 50 + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78e8dad0223f4c40",
   "metadata": {},
   "source": [
    "## Run KALE on multiple datasets"
   ]
  },
  {
   "cell_type": "code",
   "id": "385fc916a7f1f872",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:31:10.991350Z",
     "start_time": "2025-11-19T20:30:47.662603Z"
    }
   },
   "source": [
    "# Datasets which has MYC perturbations\n",
    "adata_files = [\n",
    "    \"data/scPerturbData/FrangiehIzar2021_RNA.h5ad\",\n",
    "    \"data/scPerturbData/ReplogleWeissman2022_K562_essential.h5ad\",\n",
    "    \"data/scPerturbData/ReplogleWeissman2022_rpe1.h5ad\",\n",
    "]\n",
    "\n",
    "for adata_file in adata_files:\n",
    "    if not os.path.exists(adata_file):\n",
    "        print(f\"\\nFile {adata_file} does not exist. Skipping...\\n\")\n",
    "        continue\n",
    "\n",
    "    print(\"Running KALE for test case: \", adata_file)\n",
    "\n",
    "    file_name = adata_file.split('/')[-1].replace('.h5ad', '')\n",
    "\n",
    "    # run script\n",
    "    print(\"Running KALE for test case without weights without ignore_zeros\")\n",
    "    !uv run src/kale.py --gene_exp_file {adata_file} --prior_file data/causal_priors.tsv --output_file data/_kale_scores_unweighted_{file_name}.tsv --pvalue_output_file data/_kale_pvalues_unweighted_{file_name}.tsv --ignore_zeros False --cores 8 --method rank_of_ranks --min_targets 0 --weighted False\n",
    "\n",
    "    print(\"Running KALE for test case without weights ignoring zeros\")\n",
    "    !uv run src/kale.py --gene_exp_file {adata_file} --prior_file data/causal_priors.tsv --output_file data/_kale_scores_unweighted_ignore_zeros_{file_name}.tsv --pvalue_output_file data/_kale_pvalues_unweighted_ignore_zeros_{file_name}.tsv --ignore_zeros True --cores 8 --method rank_of_ranks --min_targets 1 --weighted False\n",
    "\n",
    "\n",
    "    print(\"Running KALE for test case with weights without ignore_zeros\")\n",
    "    !uv run src/kale.py --gene_exp_file {adata_file} --prior_file data/causal_priors.tsv --output_file data/_kale_scores_weighted_{file_name}.tsv --pvalue_output_file data/_kale_pvalues_weighted_{file_name}.tsv --ignore_zeros False --cores 8 --method rank_of_ranks --min_targets 1 --weighted True --weighted_power_factor 1\n",
    "\n",
    "\n",
    "    print(\"Running KALE for test case with weights ignoring zeros\")\n",
    "    !uv run src/kale.py --gene_exp_file {adata_file} --prior_file data/causal_priors.tsv --output_file data/_kale_scores_weighted_ignore_zeros_{file_name}.tsv --pvalue_output_file data/_kale_pvalues_weighted_ignore_zeros_{file_name}.tsv --ignore_zeros True --cores 8 --method rank_of_ranks --min_targets 1 --weighted True --weighted_power_factor 1"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running KALE for test case:  data/scPerturbData/AdamsonWeissman2016_GSM2406675_10X001_kt_filtered.h5ad\n",
      "Running KALE for test case without weights without ignore_zeros\n",
      "Preprocessing gene expression data...\r\n",
      "Using raw gene expression as input for per-cell ranking...\r\n",
      "Starting TF activity using 8 cores.\r\n",
      "Running in parallel with CORES_USED=8.\r\n",
      "Processing cells in parallel:   0%|                    | 0/3999 [00:00<?, ?it/s][Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.\r\n",
      "Processing cells in parallel:   1%|           | 32/3999 [00:01<01:53, 34.97it/s][Parallel(n_jobs=8)]: Done   34 out of 3999 | elapsed:    1.1s\r\n",
      "Processing cells in parallel:  45%|███▌    | 1808/3999 [00:03<00:03, 650.34it/s][Parallel(n_jobs=8)]: Done 1680 out of 3999 | elapsed:    3.8s\r\n",
      "Processing cells in parallel: 100%|████████| 3999/3999 [00:07<00:00, 556.32it/s]\r\n",
      "[Parallel(n_jobs=8)]: Done 3999 out of 3999 | elapsed:    7.5s finished\r\n",
      "\r\n",
      "Aggregating results...\r\n",
      "kale completed\r\n",
      "Kale TF activity scores and p-values have been saved.\r\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "id": "a5c4d34a3917b5ab",
   "metadata": {},
   "source": [
    "## Run Viper, MLM, ULM, "
   ]
  },
  {
   "cell_type": "code",
   "id": "32aa357fa2d6a058",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:57.402920Z",
     "start_time": "2025-11-19T20:31:10.997262Z"
    }
   },
   "source": [
    "# Datasets which has MYC perturbations\n",
    "adata_files = [\n",
    "    \"data/scPerturbData/FrangiehIzar2021_RNA.h5ad\",\n",
    "    \"data/scPerturbData/ReplogleWeissman2022_K562_essential.h5ad\",\n",
    "    \"data/scPerturbData/ReplogleWeissman2022_rpe1.h5ad\",\n",
    "]\n",
    "\n",
    "net_file = \"data/causal_priors.tsv\"\n",
    "effect_map = {\"upregulates-expression\": 1, \"downregulates-expression\": -1}\n",
    "net = pd.read_csv(\n",
    "    net_file,\n",
    "    sep=\"\\t\",\n",
    "    names=[\"source\", \"weight\", \"target\"],\n",
    "    usecols=[0, 1, 2],\n",
    "    converters={\"weight\": effect_map.get}\n",
    ")[[\"source\", \"target\", \"weight\"]]\n",
    "\n",
    "# Run Decoupler Methods\n",
    "# methods_to_run = [\"viper\", \"mlm\", \"ulm\"]\n",
    "methods_to_run = [\"viper\"]\n",
    "\n",
    "for adata_file in adata_files:\n",
    "    print(f\"Processing {adata_file}...\")\n",
    "    adata = sc.read_h5ad(adata_file)\n",
    "\n",
    "    # Filter low-quality genes and cells\n",
    "    sc.pp.filter_cells(adata, min_genes=500)\n",
    "    sc.pp.filter_genes(adata, min_cells=10)\n",
    "    print(\"Shape after filtering: \", adata.shape)\n",
    "\n",
    "    # Mitochondrial/QC filter\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('MT-')  # assuming human data; use 'mt-' for mouse\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "    adata = adata[adata.obs.pct_counts_mt < 20, :]  # filter cells with >20% mitochondrial genes\n",
    "    print(\"Final Shape \", adata.shape)\n",
    "\n",
    "    print(\"Normalizing and Log1p...\")\n",
    "    # 1. Normalize (Corrects sequencing depth)\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    # 2. Log transform (Stabilizes variance / makes data Gaussian-like)\n",
    "    sc.pp.log1p(adata)\n",
    "    # 3. Scale (Optional for ULM/MLM, Highly Recommended for VIPER)\n",
    "    sc.pp.scale(adata, max_value=10)\n",
    "\n",
    "    dc.mt.decouple(adata, net, tmin=0, methods=methods_to_run)\n",
    "\n",
    "    for method in methods_to_run:\n",
    "        method_scores = adata.obsm[f\"score_{method}\"]\n",
    "        method_scores_df = pd.DataFrame(\n",
    "            method_scores,\n",
    "            index=adata.obs_names,\n",
    "            columns=[col for col in adata.var_names if col in net['source'].values]\n",
    "        )\n",
    "        method_scores_df.to_csv(f\"data/_{method}_scores_{adata_file.split('/')[-1].replace('.h5ad', '')}.tsv\", sep=\"\\t\")\n",
    "        print(f\"Saved {method} scores to data/_{method}_scores_{adata_file.split('/')[-1].replace('.h5ad', '')}.tsv\")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing data/scPerturbData/AdamsonWeissman2016_GSM2406675_10X001_kt_filtered.h5ad...\n",
      "Normalizing and Log1p...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/functools.py:907: UserWarning: zero-centering a sparse array/matrix densifies it.\n",
      "  return dispatch(args[0].__class__)(*args, **kw)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved viper scores to data/_viper_scores_AdamsonWeissman2016_GSM2406675_10X001_kt_filtered.tsv\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:58.149650Z",
     "start_time": "2025-11-19T20:32:57.493425Z"
    }
   },
   "cell_type": "code",
   "source": [
    "viper_scores = pd.read_csv(\"data/_viper_scores_AdamsonWeissman2016_GSM2406675_10X001_kt_filtered.tsv\", sep=\"\\t\",\n",
    "                           index_col=0)\n",
    "# viper_scores = viper_scores.dropna(axis=1, how='all')\n",
    "viper_scores = viper_scores.reindex(sorted(viper_scores.columns), axis=1)\n",
    "viper_scores.head()"
   ],
   "id": "746d4635a045f5b5",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                A1BG      AATF     ABCA3      ABL1     ACACA    ACADSB  \\\n",
       "cell_barcode                                                             \n",
       "AAACATACACCGAT   NaN -1.617859 -0.744422 -0.328945 -0.550234  0.078765   \n",
       "AAACATACAGAGAT   NaN  1.312137  0.631046 -0.338817 -0.538363  0.089844   \n",
       "AAACATACGTTGAC   NaN -0.679692  0.804137  1.475622 -0.635682  0.009033   \n",
       "AAACCGTGCAGCTA   NaN  2.461895 -0.817146 -0.616701 -0.585374  0.050001   \n",
       "AAACCGTGGAACTC   NaN  0.187228 -0.363359  0.206227 -0.370617  0.260234   \n",
       "\n",
       "                   ACAT2      ACLY    ACTL6A     ADCY7  ...   ZNF385A  ZNF420  \\\n",
       "cell_barcode                                            ...                     \n",
       "AAACATACACCGAT -0.550234 -1.178748  0.909472 -0.049800  ...  0.673482     NaN   \n",
       "AAACATACAGAGAT -0.538363 -1.153887 -2.116118 -0.038749  ...  0.691471     NaN   \n",
       "AAACATACGTTGAC -0.635682 -1.320915 -0.596403 -0.121333  ...  0.593733     NaN   \n",
       "AAACCGTGCAGCTA -0.585374  0.662877  0.122142 -0.078161  ...  0.639863     NaN   \n",
       "AAACCGTGGAACTC -0.370617 -0.902204  2.456166  0.127604  ...  0.920516     NaN   \n",
       "\n",
       "                ZNF444  ZNF467  ZNF521  ZNF638     ZNF76     ZNFX1     ZNRD1  \\\n",
       "cell_barcode                                                                   \n",
       "AAACATACACCGAT     NaN     NaN     NaN     NaN -1.098185 -0.005620  0.715378   \n",
       "AAACATACAGAGAT     NaN     NaN     NaN     NaN -1.102796 -0.014653  0.734434   \n",
       "AAACATACGTTGAC     NaN     NaN     NaN     NaN -1.115631  0.065684  0.633717   \n",
       "AAACCGTGCAGCTA     NaN     NaN     NaN     NaN -1.105961  0.023487  0.681056   \n",
       "AAACCGTGGAACTC     NaN     NaN     NaN     NaN -1.138586 -0.183890  0.971382   \n",
       "\n",
       "                     ZYX  \n",
       "cell_barcode              \n",
       "AAACATACACCGAT -0.628818  \n",
       "AAACATACAGAGAT -0.619064  \n",
       "AAACATACGTTGAC  1.847311  \n",
       "AAACCGTGCAGCTA -0.667702  \n",
       "AAACCGTGGAACTC -0.449006  \n",
       "\n",
       "[5 rows x 1291 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>AATF</th>\n",
       "      <th>ABCA3</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACACA</th>\n",
       "      <th>ACADSB</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>ACLY</th>\n",
       "      <th>ACTL6A</th>\n",
       "      <th>ADCY7</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF385A</th>\n",
       "      <th>ZNF420</th>\n",
       "      <th>ZNF444</th>\n",
       "      <th>ZNF467</th>\n",
       "      <th>ZNF521</th>\n",
       "      <th>ZNF638</th>\n",
       "      <th>ZNF76</th>\n",
       "      <th>ZNFX1</th>\n",
       "      <th>ZNRD1</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACACCGAT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.617859</td>\n",
       "      <td>-0.744422</td>\n",
       "      <td>-0.328945</td>\n",
       "      <td>-0.550234</td>\n",
       "      <td>0.078765</td>\n",
       "      <td>-0.550234</td>\n",
       "      <td>-1.178748</td>\n",
       "      <td>0.909472</td>\n",
       "      <td>-0.049800</td>\n",
       "      <td>...</td>\n",
       "      <td>0.673482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.098185</td>\n",
       "      <td>-0.005620</td>\n",
       "      <td>0.715378</td>\n",
       "      <td>-0.628818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACAGAGAT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.312137</td>\n",
       "      <td>0.631046</td>\n",
       "      <td>-0.338817</td>\n",
       "      <td>-0.538363</td>\n",
       "      <td>0.089844</td>\n",
       "      <td>-0.538363</td>\n",
       "      <td>-1.153887</td>\n",
       "      <td>-2.116118</td>\n",
       "      <td>-0.038749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.691471</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.102796</td>\n",
       "      <td>-0.014653</td>\n",
       "      <td>0.734434</td>\n",
       "      <td>-0.619064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACGTTGAC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.679692</td>\n",
       "      <td>0.804137</td>\n",
       "      <td>1.475622</td>\n",
       "      <td>-0.635682</td>\n",
       "      <td>0.009033</td>\n",
       "      <td>-0.635682</td>\n",
       "      <td>-1.320915</td>\n",
       "      <td>-0.596403</td>\n",
       "      <td>-0.121333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.593733</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.115631</td>\n",
       "      <td>0.065684</td>\n",
       "      <td>0.633717</td>\n",
       "      <td>1.847311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCAGCTA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.461895</td>\n",
       "      <td>-0.817146</td>\n",
       "      <td>-0.616701</td>\n",
       "      <td>-0.585374</td>\n",
       "      <td>0.050001</td>\n",
       "      <td>-0.585374</td>\n",
       "      <td>0.662877</td>\n",
       "      <td>0.122142</td>\n",
       "      <td>-0.078161</td>\n",
       "      <td>...</td>\n",
       "      <td>0.639863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.105961</td>\n",
       "      <td>0.023487</td>\n",
       "      <td>0.681056</td>\n",
       "      <td>-0.667702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGGAACTC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.187228</td>\n",
       "      <td>-0.363359</td>\n",
       "      <td>0.206227</td>\n",
       "      <td>-0.370617</td>\n",
       "      <td>0.260234</td>\n",
       "      <td>-0.370617</td>\n",
       "      <td>-0.902204</td>\n",
       "      <td>2.456166</td>\n",
       "      <td>0.127604</td>\n",
       "      <td>...</td>\n",
       "      <td>0.920516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.138586</td>\n",
       "      <td>-0.183890</td>\n",
       "      <td>0.971382</td>\n",
       "      <td>-0.449006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1291 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:58.915700Z",
     "start_time": "2025-11-19T20:32:58.248439Z"
    }
   },
   "cell_type": "code",
   "source": [
    "kale_scores = pd.read_csv(\"data/_kale_scores_unweighted_AdamsonWeissman2016_GSM2406675_10X001_kt_filtered.tsv\",\n",
    "                          sep=\"\\t\", index_col=0)\n",
    "# kale_scores = kale_scores.dropna(axis=1, how='all')\n",
    "kale_scores = kale_scores.reindex(sorted(kale_scores.columns), axis=1)\n",
    "kale_scores.head()"
   ],
   "id": "a4ecf8949f7f6418",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                     A2M      AATF     ABCA1     ABCA3     ABCB1     ABCG1  \\\n",
       "cell_barcode                                                                 \n",
       "AAACATACACCGAT -1.831116 -1.509827  0.034511 -0.811318  1.106163 -0.895033   \n",
       "AAACATACAGAGAT  1.887936  0.668283  2.119007  0.676800 -0.408522  1.210134   \n",
       "AAACATACGTTGAC  1.664323  0.473649 -0.002047  0.654262 -0.943500 -1.037380   \n",
       "AAACCGTGCAGCTA  1.777362  3.766162 -2.117577 -0.877021 -0.852840 -0.942169   \n",
       "AAACCGTGGAACTC -1.607690  0.269451  0.156968 -0.331397 -0.108344 -0.529982   \n",
       "\n",
       "                   ABCG5     ABCG8      ABL1     ACACA  ...     ZNF24  \\\n",
       "cell_barcode                                            ...             \n",
       "AAACATACACCGAT -0.895033 -0.895033 -0.216546 -0.719286  ... -0.784743   \n",
       "AAACATACAGAGAT  1.210134  1.210134 -0.137432 -0.686871  ... -0.782145   \n",
       "AAACATACGTTGAC -1.037380 -1.037380  0.959447 -0.811102  ...  1.092437   \n",
       "AAACCGTGCAGCTA -0.942169 -0.942169 -0.285025 -0.742611  ... -0.780199   \n",
       "AAACCGTGGAACTC -0.529982 -0.529982  0.175601 -0.465175  ... -0.810883   \n",
       "\n",
       "                  ZNF300    ZNF350    ZNF382    ZNF383   ZNF385A     ZNF76  \\\n",
       "cell_barcode                                                                 \n",
       "AAACATACACCGAT -1.645247 -0.137264  1.743215  0.510729  0.940569 -1.724501   \n",
       "AAACATACAGAGAT -1.291756 -0.137973  1.198376  0.486617  0.969748 -1.726453   \n",
       "AAACATACGTTGAC  0.462682 -0.135670  1.703406  0.596243  0.835394 -1.712455   \n",
       "AAACCGTGCAGCTA  0.531486 -2.927052 -0.475886  0.529959  0.901276 -1.720245   \n",
       "AAACCGTGGAACTC  0.219799 -0.151256  0.982213  0.276617  1.291450 -1.794669   \n",
       "\n",
       "                   ZNFX1     ZNRD1       ZYX  \n",
       "cell_barcode                                  \n",
       "AAACATACACCGAT -0.026916  1.030724 -0.830515  \n",
       "AAACATACAGAGAT -0.043292  1.060918 -0.799975  \n",
       "AAACATACGTTGAC  0.033070  0.922208  2.131584  \n",
       "AAACCGTGCAGCTA -0.005994  0.989651 -0.859558  \n",
       "AAACCGTGGAACTC -0.230872  1.392857 -0.573505  \n",
       "\n",
       "[5 rows x 1572 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M</th>\n",
       "      <th>AATF</th>\n",
       "      <th>ABCA1</th>\n",
       "      <th>ABCA3</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCG1</th>\n",
       "      <th>ABCG5</th>\n",
       "      <th>ABCG8</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACACA</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF24</th>\n",
       "      <th>ZNF300</th>\n",
       "      <th>ZNF350</th>\n",
       "      <th>ZNF382</th>\n",
       "      <th>ZNF383</th>\n",
       "      <th>ZNF385A</th>\n",
       "      <th>ZNF76</th>\n",
       "      <th>ZNFX1</th>\n",
       "      <th>ZNRD1</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACACCGAT</th>\n",
       "      <td>-1.831116</td>\n",
       "      <td>-1.509827</td>\n",
       "      <td>0.034511</td>\n",
       "      <td>-0.811318</td>\n",
       "      <td>1.106163</td>\n",
       "      <td>-0.895033</td>\n",
       "      <td>-0.895033</td>\n",
       "      <td>-0.895033</td>\n",
       "      <td>-0.216546</td>\n",
       "      <td>-0.719286</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.784743</td>\n",
       "      <td>-1.645247</td>\n",
       "      <td>-0.137264</td>\n",
       "      <td>1.743215</td>\n",
       "      <td>0.510729</td>\n",
       "      <td>0.940569</td>\n",
       "      <td>-1.724501</td>\n",
       "      <td>-0.026916</td>\n",
       "      <td>1.030724</td>\n",
       "      <td>-0.830515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACAGAGAT</th>\n",
       "      <td>1.887936</td>\n",
       "      <td>0.668283</td>\n",
       "      <td>2.119007</td>\n",
       "      <td>0.676800</td>\n",
       "      <td>-0.408522</td>\n",
       "      <td>1.210134</td>\n",
       "      <td>1.210134</td>\n",
       "      <td>1.210134</td>\n",
       "      <td>-0.137432</td>\n",
       "      <td>-0.686871</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.782145</td>\n",
       "      <td>-1.291756</td>\n",
       "      <td>-0.137973</td>\n",
       "      <td>1.198376</td>\n",
       "      <td>0.486617</td>\n",
       "      <td>0.969748</td>\n",
       "      <td>-1.726453</td>\n",
       "      <td>-0.043292</td>\n",
       "      <td>1.060918</td>\n",
       "      <td>-0.799975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACGTTGAC</th>\n",
       "      <td>1.664323</td>\n",
       "      <td>0.473649</td>\n",
       "      <td>-0.002047</td>\n",
       "      <td>0.654262</td>\n",
       "      <td>-0.943500</td>\n",
       "      <td>-1.037380</td>\n",
       "      <td>-1.037380</td>\n",
       "      <td>-1.037380</td>\n",
       "      <td>0.959447</td>\n",
       "      <td>-0.811102</td>\n",
       "      <td>...</td>\n",
       "      <td>1.092437</td>\n",
       "      <td>0.462682</td>\n",
       "      <td>-0.135670</td>\n",
       "      <td>1.703406</td>\n",
       "      <td>0.596243</td>\n",
       "      <td>0.835394</td>\n",
       "      <td>-1.712455</td>\n",
       "      <td>0.033070</td>\n",
       "      <td>0.922208</td>\n",
       "      <td>2.131584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCAGCTA</th>\n",
       "      <td>1.777362</td>\n",
       "      <td>3.766162</td>\n",
       "      <td>-2.117577</td>\n",
       "      <td>-0.877021</td>\n",
       "      <td>-0.852840</td>\n",
       "      <td>-0.942169</td>\n",
       "      <td>-0.942169</td>\n",
       "      <td>-0.942169</td>\n",
       "      <td>-0.285025</td>\n",
       "      <td>-0.742611</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.780199</td>\n",
       "      <td>0.531486</td>\n",
       "      <td>-2.927052</td>\n",
       "      <td>-0.475886</td>\n",
       "      <td>0.529959</td>\n",
       "      <td>0.901276</td>\n",
       "      <td>-1.720245</td>\n",
       "      <td>-0.005994</td>\n",
       "      <td>0.989651</td>\n",
       "      <td>-0.859558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGGAACTC</th>\n",
       "      <td>-1.607690</td>\n",
       "      <td>0.269451</td>\n",
       "      <td>0.156968</td>\n",
       "      <td>-0.331397</td>\n",
       "      <td>-0.108344</td>\n",
       "      <td>-0.529982</td>\n",
       "      <td>-0.529982</td>\n",
       "      <td>-0.529982</td>\n",
       "      <td>0.175601</td>\n",
       "      <td>-0.465175</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.810883</td>\n",
       "      <td>0.219799</td>\n",
       "      <td>-0.151256</td>\n",
       "      <td>0.982213</td>\n",
       "      <td>0.276617</td>\n",
       "      <td>1.291450</td>\n",
       "      <td>-1.794669</td>\n",
       "      <td>-0.230872</td>\n",
       "      <td>1.392857</td>\n",
       "      <td>-0.573505</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1572 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:59.553419Z",
     "start_time": "2025-11-19T20:32:59.550603Z"
    }
   },
   "cell_type": "code",
   "source": [
    "intersec = set(kale_scores.columns).intersection(set(viper_scores.columns))\n",
    "print(f\"Number of common TFs between KALE and VIPER: {len(intersec)}\")"
   ],
   "id": "de4e1635f2c1574d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of common TFs between KALE and VIPER: 1070\n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:59.608467Z",
     "start_time": "2025-11-19T20:32:59.593806Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# remove rows if net[\"target\"] not in adata.var_names\n",
    "print(\"net shape before filtering: \", net.shape)\n",
    "net = net[net['target'].isin(adata.var_names)]\n",
    "print(\"net shape after filtering: \", net.shape)\n",
    "# net_grouped_targets = filtered_net.groupby('source')['target'].apply(list)\n",
    "# net_grouped_targets = net_grouped_targets.reset_index()\n",
    "# net_grouped_targets.columns = ['source', 'targets']\n",
    "# net_grouped_targets = net_grouped_targets.set_index('source')\n",
    "# net_grouped_targets"
   ],
   "id": "44addabe7bd5e1e3",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "net shape before filtering:  (11839, 3)\n",
      "net shape after filtering:  (6943, 3)\n"
     ]
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:59.627917Z",
     "start_time": "2025-11-19T20:32:59.618868Z"
    }
   },
   "cell_type": "code",
   "source": [
    "net_grouped = net.groupby('source').size().sort_values(ascending=False)\n",
    "net_grouped = net_grouped.reset_index()\n",
    "net_grouped.columns = ['source', 'target_count']\n",
    "net_grouped = net_grouped.set_index('source')\n",
    "print(\"Kale and network intersection:\", len(set(kale_scores.columns).intersection(set(net_grouped.index))))\n",
    "print(\"Viper and network intersection:\", len(set(viper_scores.columns).intersection(set(net_grouped.index))))"
   ],
   "id": "f08667ffda0d9672",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kale and network intersection: 1572\n",
      "Viper and network intersection: 1070\n"
     ]
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:59.745099Z",
     "start_time": "2025-11-19T20:32:59.741513Z"
    }
   },
   "cell_type": "code",
   "source": [
    "all_genes = set(adata.var.index.tolist())\n",
    "print(\"Total genes in dataset:\", len(all_genes))\n",
    "print(\"Total TFs in network:\", len(net_grouped))\n",
    "print(\"Total TFs in KALE scores:\", len(kale_scores.columns))\n",
    "print(\"Total TFs in VIPER scores:\", len(viper_scores.columns))\n",
    "print(\"Intersection of total genes(in datasets) with network TFs: \",\n",
    "      len(all_genes.intersection(set(net_grouped.index.tolist()))))\n",
    "print(\"Intersection of total genes(in datasets) with with KALE TFs: \",\n",
    "      len(all_genes.intersection(set(kale_scores.columns.tolist()))))\n",
    "print(\"Intersection of total genes(in datasets) with with VIPER TFs: \",\n",
    "      len(all_genes.intersection(set(viper_scores.columns.tolist()))))"
   ],
   "id": "7547494ded1692f1",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total genes in dataset: 12487\n",
      "Total TFs in network: 1572\n",
      "Total TFs in KALE scores: 1572\n",
      "Total TFs in VIPER scores: 1291\n",
      "Intersection of total genes(in datasets) with network TFs:  1070\n",
      "Intersection of total genes(in datasets) with with KALE TFs:  1070\n",
      "Intersection of total genes(in datasets) with with VIPER TFs:  1291\n"
     ]
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-19T20:32:59.820189Z",
     "start_time": "2025-11-19T20:32:59.800945Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# group net by making targets list\n",
    "net_grouped_targets = net.groupby('source')['target'].apply(list)\n",
    "net_grouped_targets = net_grouped_targets.reset_index()\n",
    "net_grouped_targets.columns = ['source', 'targets']\n",
    "net_grouped_targets = net_grouped_targets.set_index('source')\n",
    "net_grouped_targets.head()"
   ],
   "id": "d9268cf41db362e6",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                  targets\n",
       "source                                                   \n",
       "A2M                                               [STAT3]\n",
       "AATF                                   [BAX, CTNNB1, MYC]\n",
       "ABCA1                              [NR1H2, NR1H3, SREBF2]\n",
       "ABCA3                      [SREBF1, GATA6, NFATC3, CEBPA]\n",
       "ABCB1   [SP1, RELB, RELA, TCF7L2, FOXO1, EGR1, CEBPB, ..."
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targets</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>[STAT3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATF</th>\n",
       "      <td>[BAX, CTNNB1, MYC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA1</th>\n",
       "      <td>[NR1H2, NR1H3, SREBF2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA3</th>\n",
       "      <td>[SREBF1, GATA6, NFATC3, CEBPA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB1</th>\n",
       "      <td>[SP1, RELB, RELA, TCF7L2, FOXO1, EGR1, CEBPB, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scDataStatProject",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
